### PR TITLE
added memory layout

### PR DESCRIPTION
**Pull request by ntasfi**

### CODE REVIEW

File: memory.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        max_size: int
+            The max amount of samples we want to hold until we start 
+            deleting the oldest sample
+
+        observation_size: tuple
+            A tuple describing a single observations size from the environment. 
+            Does NOT include the batch size.
+            This module will try to be agnostic to the size pass as long as its 
+            an image.
+            eg. (84, 84, 3) or (3, 84, 84) or (84,84)
```
Comment (EderSantana) [inline memory.py:16]: will we pick one or let it open? I mean between r-c-c and c-r-c?

File: memory.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        #return the length of the memory
+        pass
+
+    def _adjust_memory(self):
+        """
+            this will take care of adjusting the memory size so it states within the limits
+            we can be lazy with this and only adjust memory when we train batches
+            we would overflow a little bit not a big deal.
+        """
+        pass
```
Comment (EderSantana) [inline memory.py:33]: I didn't get where will this be used

File: memory.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    def add(self, obs, action, reward, terminal):
+        """
+            Check is obs is the same as observation_size
+            Add to the last spot in memory (as a tuple of the above)
+            Adjust memory size if its over the max_size
+
+            obs will be stored as int8
+            action will be stored as int8
+            reward will be stored as float32
+            terminal will be stored as int8 or bool (@eder what do you think)
```
Comment (EderSantana) [inline memory.py]: either one numpy knows how to handle in an equation, I've been using bool

File: memory.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        pass
+
+    def random_batch(self, batch_size, state_size):
```
Comment (EderSantana) [inline memory.py]: `observation_size` and `state_size` have different meanings, but similar names. I'd suggest `state_length` no, biggy though

File: memory.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        """
+            state_size is how many single observations constitue a single 'state'
+            eg. DQN usually uses 4 observations to represent a single state.
+
+            batch_size is the size of the batch we return back. Will have shape:
+            (batch_size, state_size, observation_size)
+            eg. (32, 4, 84, 84) or (32, 4, 3, 16, 16) or (32, 4, 16, 16, 3) etc.
```
Comment (EderSantana) [inline memory.py:62]: I was doing that, but most of the time for theano and tensorflow just concatenating everything in `axis=1` is just fine. `axis=1` is channel for images, time for sequences and dim for vectors, those are usually always the dimensions we concat. I'd just do that job for the user already. Probably a freedom we won't miss... I might be wrong, did you have anything else in mind?

File: memory.py (hunk 1 above)
Comment (ntasfi) [inline memory.py:16]: we can keep it open, it shouldn't matter as long as the tuple for the format is passed in! When we are doing things like resizing the image we could assume the smallest value (if there are more than 2) is the channel.

File: memory.py (hunk 2 above)
Comment (ntasfi) [inline memory.py:33]: It would be used internally by the Memory module

File: memory.py (hunk 3 above)
Comment (ntasfi) [inline memory.py]: kk we can use int8 perhaps to keep it numeric?

File: memory.py (hunk 4 above)
Comment (ntasfi) [inline memory.py]: I could rename it to `observation_dim` and `state_length`?

File: memory.py (hunk 5 above)
Comment (ntasfi) [inline memory.py:62]: Yea lets keep it similar to how others have it then, no need to deviate!

File: memory.py (hunk 3 above)
Comment (EderSantana) [inline memory.py]: I think that is fine

File: memory.py (hunk 4 above)
Comment (EderSantana) [inline memory.py]: keras standard is to call it shape actually. so it would be `observation_shape` and `state_length`

**Comment by EderSantana** [comment]
btw, do we need a `get_current` or `get_last` to pick the input for the current action?

File: memory.py (hunk 4 above)
Comment (ntasfi) [inline memory.py]: lets go with keras standard.

**Comment by ntasfi** [comment]
Hmm, we could do that on the agent. I think it might be good to keep it in the memory just so it has a clear separation of responsibilities

**Comment by EderSantana** [comment]
> we could do that on the agent

great! so the agent should have a current state that is a deque?

**Comment by ntasfi** [comment]
@EderSantana Sorry I became very busy the last week! 

Yes we can do that or have the memory class provide a function that gets the last N observations!
